Plugins/NativeCapture/Source/NativeCapture.cpp

```diff
@@ -0,0 +1,81 @@
+#include "NativeCapture.h"
+
+#include <Babylon/JsRuntime.h>
+#include <GraphicsImpl.h>
+
+#include <vector>
+
+namespace Babylon::Plugins::Internal
+{
+    class NativeCapture : public Napi::ObjectWrap<NativeCapture>
+    {
+    public:
+        static constexpr auto JS_CLASS_NAME = "NativeCapture";
+
+        static void Initialize(Napi::Env env)
+        {
+            Napi::HandleScope scope{env};
+
+            Napi::Function func = NativeCapture::DefineClass(
+                env,
+                JS_CLASS_NAME,
+                {
+                    NativeCapture::InstanceMethod("addOnCaptureCallback", &NativeCapture::AddOnCaptureCallback),
+                    NativeCapture::InstanceMethod("dispose", &NativeCapture::Dispose),
+                });
+
+            env.Global().Set(JS_CLASS_NAME, func);
+        }
+
+        NativeCapture(const Napi::CallbackInfo& info)
+            : Napi::ObjectWrap<NativeCapture>{info}
+            , m_runtime{JsRuntime::GetFromJavaScript(info.Env())}
+            , m_graphicsImpl(Graphics::Impl::GetFromJavaScript(info.Env()))
+            , m_ticket{m_graphicsImpl.AddCaptureCallback([this](auto& data) { CaptureDataReceived(data); })}
+        {
+            m_graphicsImpl.StartCapture();
+        }
+
+    private:
+        void AddOnCaptureCallback(const Napi::CallbackInfo& info)
+        {
+            auto listener = info[0].As<Napi::Function>();
+            m_callbacks.push_back(Napi::Persistent(listener));
+        }
+
+        void CaptureDataReceived(const BgfxCallback::CaptureData& data)
+        {
+            std::vector<uint8_t> bytes{};
+            bytes.resize(data.DataSize);
+            std::memcpy(bytes.data(), data.Data, data.DataSize);
+            m_runtime.Dispatch([this, data = data, bytes = std::move(bytes)](Napi::Env env) mutable {
+                data.Data = bytes.data();
+
+                auto external = Napi::External<BgfxCallback::CaptureData>::New(env, &data);
+                for (const auto& callback : m_callbacks)
+                {
+                    callback.Call({external});
```

Comment:
Isn't this basically passing an opaque object to the JS side? We need the JS side to be able to read the actual bytes (or pass it along to other native code that doesn't know anything about Napi or Babylon Native). I was expecting we'd pass something like a `Uint8Array` and a width/stride.

Reply:
Spoke offline about this. It is passing an opaque object with the current expectation that that object would be interpreted by other native objects for maximum efficiency. We also discussed a mechanism to "un-opaque" this object if JS wants to read it directly; is that something we currently have a use case for such that we should put it in right now, or should we add that later if/as it becomes necessary?

Reply:
The JS does not need a readable but array, but it needs a pointer that can be passed to C++ code (via JSI).

Reply:
In the latest iteration, I modified the callback to take a JS object that automatically exposes the callback data, including an ArrayBuffer for the bytes. This ArrayBuffer is made using N-API's external ArrayBuffer mechanism, which does not take ownership of the memory. This implementation balances what I recall to be the results from our conversation about this a few weeks ago, but it definitely has some tradeoffs that I'll be happy to discuss and revise. (For example, I'm not a huge fan of `new`ing a N-API object for every callback, but since caching had pave-over issues I'm not sure if another pattern would be better. It could certainly be more efficient if I just pass up a raw pointer, but that consideration must be balanced against safety/clarity/ease of use for the relevant scenario, so I'm interested in perspectives on that.)